File: src/VisualStudio/Reqnroll.IdeSupport.VisualStudio.VSSDKIntegration/TestReporter/MtpEphemeralInjection.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Reqnroll.IdeSupport.Common.Logging;

namespace Reqnroll.IdeSupport.VisualStudio.TestReporter;

/// <summary>
/// Ephemerally compiles the bundled <c>Reqnroll.IdeSupport.TestReporter.MTP</c> into every
/// Microsoft.Testing.Platform (MTP)-capable project built in this <c>devenv.exe</c> session — issue
/// #715 plan §5.6 — via the <c>CustomAfterMicrosoftCommonTargets</c> MSBuild extensibility point,
/// instead of the VSTest logger's runsettings-based per-run injection (<c>ReqnrollTestLoggerRunSettingsService</c>),
/// which under MTP is either silently ignored or, in the native <c>dotnet test</c> mode, not
/// applicable at all (VS's own Test Explorer never shells out to <c>dotnet test</c> in the first
/// place — plan §5.7: MTP-capable projects always run through VS's own "testing platform server
/// mode").
/// </summary>
/// <remarks>
/// <para>
/// <b>Set once per session, not per run</b> — unlike Rider's <c>ProcessBuilder</c>-per-invocation
/// model (issue #715 phase 4, Rider leg), VS's Test Explorer builds are not individually controlled
/// by this extension, so the only lever available is a process-wide environment variable, set as
/// early as possible (package load) so it's in place before any build happens in this session — same
/// reasoning as the existing autoload pattern (<c>ProvideAutoLoadAttribute</c> on
/// <c>ReqnrollPluginPackage</c>).
/// </para>
/// <para>
/// <b>Solution-wide gating.</b> The injected <c>&lt;Reference&gt;</c> is harmless but wasteful for a
/// project that doesn't use it, so this only sets the environment variable at all when at least one
/// project under the solution looks MTP-capable (<see cref="MtpProjectDetection"/>) — the common
/// all-VSTest solution never pays for it. A mixed solution still injects the reference into every
/// project built in this session, VSTest-only siblings included — an accepted, documented
/// imprecision (plan §5.7's "mixed solutions fall out for free" is about per-run detection like
/// Rider's/VS Code's; VS's coarser env-var-per-session model can't scope this exactly without giving
/// it a way to intercept individual Test Explorer builds, which it does not have).
/// </para>
/// <para>
/// <b>Unverified (plan §7 risk #1)</b>: whether VS's in-process MSBuild evaluation honors a
/// process-level environment variable set by the VSIX at load time is not live-tested. If it doesn't,
/// this degrades to exactly today's shipped behavior for MTP-capable projects (the pre-#700
/// reflection-bridge fallback) — not a regression, just no improvement, same framing the plan uses
/// throughout for this risk.
/// </para>
/// </remarks>
public static class MtpEphemeralInjection
{
    /// <summary>The MSBuild extensibility property this mechanism relies on (plan §5.6).</summary>
    public const string CustomAfterMicrosoftCommonTargetsVariable = "CustomAfterMicrosoftCommonTargets";

    /// <summary>
    /// Random, permanent identifier for this hook registration (plan §5.6/§7's "Include GUID is a
    /// random identifier — never copy one from another extension's props file" note) — must match the
    /// same literal value the Rider plugin (<c>RunTestRunner.kt</c>) and the phase-2 test fixture
    /// (<c>tests/Core/TestReporterFixtures/MsTestReqnrollMtp/MsTestReqnrollMtp.Fixture.csproj</c>)
    /// declare for this same hook.
    /// </summary>
    internal const string HookGuid = "a1d3c2f0-6b8e-4f2a-9c7d-3e5f8b1a4d6c";

    private static readonly HashSet<string> ExcludedDirectoryNames =
        new(StringComparer.OrdinalIgnoreCase) { "bin", "obj", ".git", ".vs", "node_modules" };

    /// <summary>
    /// Scans <paramref name="solutionDirectory"/> for <c>.csproj</c> files (bounded — skips
    /// bin/obj/.git/.vs/node_modules) and, if any is MTP-capable, writes the ephemeral <c>.targets</c>
    /// file and sets <see cref="CustomAfterMicrosoftCommonTargetsVariable"/> for this process. No-op
    /// (env var left untouched) if <paramref name="reporterDllPath"/> isn't found, or no project is
    /// MTP-capable. Every failure is logged and swallowed — this is a best-effort enhancement, never a
    /// reason to fail package initialization.
    /// </summary>
    public static bool TryEnableForSolution(string solutionDirectory, string reporterDllPath, IIdeSupportLogger logger)
    {
        try
        {
            if (string.IsNullOrEmpty(solutionDirectory) || !Directory.Exists(solutionDirectory))
            {
                logger.LogVerbose($"{nameof(MtpEphemeralInjection)}: no solution directory to scan; not enabling.");
                return false;
            }

            if (!File.Exists(reporterDllPath))
            {
                logger.LogVerbose($"{nameof(MtpEphemeralInjection)}: bundled reporter '{reporterDllPath}' not found; not enabling.");
                return false;
            }

            var mtpCapable = EnumerateProjectFiles(solutionDirectory).Any(MtpProjectDetection.IsMtpCapable);
            if (!mtpCapable)
            {
                logger.LogVerbose($"{nameof(MtpEphemeralInjection)}: no MTP-capable project found under '{solutionDirectory}'; not enabling.");
                return false;
            }

            var preExisting = Environment.GetEnvironmentVariable(CustomAfterMicrosoftCommonTargetsVariable, EnvironmentVariableTarget.Process);
            var targetsFile = WriteTargetsFile(reporterDllPath, preExisting);
            Environment.SetEnvironmentVariable(CustomAfterMicrosoftCommonTargetsVariable, targetsFile, EnvironmentVariableTarget.Process);
            logger.LogInfo($"{nameof(MtpEphemeralInjection)}: enabled for this session — {CustomAfterMicrosoftCommonTargetsVariable}={targetsFile}");
            return true;
        }
        catch (Exception ex)
        {
            logger.LogException(ex, $"{nameof(MtpEphemeralInjection)}: TryEnableForSolution failed");
            return false;
        }
    }

    /// <summary>Bounded recursive <c>*.csproj</c> walk, pruning heavy/irrelevant directories — mirrors the Rider plugin's own ad hoc-scan philosophy.</summary>
    internal static IEnumerable<string> EnumerateProjectFiles(string root)
    {
        var stack = new Stack<string>();
        stack.Push(root);
        while (stack.Count > 0)
        {
            var dir = stack.Pop();

            IEnumerable<string> files;
            try { files = Directory.EnumerateFiles(dir, "*.csproj"); }
            catch (Exception ex) when (ex is IOException or UnauthorizedAccessException) { continue; }
            foreach (var file in files) yield return file;

            IEnumerable<string> subDirectories;
            try { subDirectories = Directory.EnumerateDirectories(dir); }
            catch (Exception ex) when (ex is IOException or UnauthorizedAccessException) { continue; }
            foreach (var subDirectory in subDirectories)
            {
                if (!ExcludedDirectoryNames.Contains(Path.GetFileName(subDirectory)))
                    stack.Push(subDirectory);
            }
        }
    }

    /// <summary>
    /// Writes a small, distinctly-named <c>.targets</c> file (plan §5.6) declaring a <c>HintPath</c>
    /// <c>&lt;Reference&gt;</c> to the bundled MTP reporter plus the
    /// <c>&lt;TestingPlatformBuilderHook&gt;</c> item that gets it auto-registered via MTP's own
    /// <c>SelfRegisteredExtensions</c> generation — never touching any project file itself.
    /// <paramref name="preExistingCustomAfterTargets"/>, when non-null, is chain-imported first (plan
    /// §7 risk #6): <c>CustomAfterMicrosoftCommonTargets</c> is a general-purpose MSBuild
    /// extensibility slot a repo could already be using for something unrelated — overwriting it
    /// outright would silently break that customization for the duration of this session.
    /// <c>Exists(...)</c> guards the import so a value that happened to be a stale/invalid path
    /// doesn't itself break every subsequent build.
    /// </summary>
    internal static string WriteTargetsFile(string reporterDllPath, string? preExistingCustomAfterTargets)
    {
        var dir = Path.Combine(Path.GetTempPath(), "reqnroll-mtp-inject-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(dir);
        var file = Path.Combine(dir, "Reqnroll.IdeSupport.TestReporter.MTP.g.targets");

        var chainImport = string.IsNullOrWhiteSpace(preExistingCustomAfterTargets)
            ? string.Empty
            : $"  <Import Project=\"{preExistingCustomAfterTargets}\" Condition=\"Exists('{preExistingCustomAfterTargets}')\" />{Environment.NewLine}";

        var xml =
            "<Project>" + Environment.NewLine +
            chainImport +
            "  <ItemGroup>" + Environment.NewLine +
            "    <Reference Include=\"Reqnroll.IdeSupport.TestReporter.MTP\">" + Environment.NewLine +
            $"      <HintPath>{reporterDllPath}</HintPath>" + Environment.NewLine +
            "    </Reference>" + Environment.NewLine +
            $"    <TestingPlatformBuilderHook Include=\"{HookGuid}\">" + Environment.NewLine +
            "      <DisplayName>Reqnroll.IdeSupport.TestReporter.MTP</DisplayName>" + Environment.NewLine +
            "      <TypeFullName>Reqnroll.IdeSupport.TestReporter.MTP.TestingPlatformBuilderHook</TypeFullName>" + Environment.NewLine +
            "    </TestingPlatformBuilderHook>" + Environment.NewLine +
            "  </ItemGroup>" + Environment.NewLine +
            "</Project>" + Environment.NewLine;

        File.WriteAllText(file, xml);
        return file;
    }
}

File: src/VSCode/src/testOutcomes/mtpEphemeralInjection.ts
import * as crypto from 'crypto';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import * as vscode from 'vscode';
import { logInfo, logWarn } from '../logging/appNotify';
import { isMtpCapable } from './mtpProjectDetection';
import { resolveMtpReporterDirectory, MTP_REPORTER_ASSEMBLY_FILE_NAME } from './mtpReporterPath';

/**
 * Ephemeral `CustomAfterMicrosoftCommonTargets` injection — issue #715 plan §5.6/§5.7, ported from
 * the Visual Studio extension's `MtpEphemeralInjection.cs` and the Rider plugin's
 * `writeEphemeralMtpTargetsFile`. Never touches any project file on disk: writes a throwaway
 * `.targets` file declaring a `<Reference>` (HintPath) to the bundled MTP reporter plus a
 * `<TestingPlatformBuilderHook>` item, which MTP's own `Microsoft.Testing.Platform.MSBuild`
 * package auto-registers for any project that imports it — every MSBuild-based build already
 * imports `CustomAfterMicrosoftCommonTargets` if it points at an existing file.
 *
 * Unlike VS's own leg (a VSIX package that owns MSBuild's own environment when it launches a
 * build), this extension has no build/test hook of its own at all: C# Dev Kit spawns `dotnet test`
 * in a child process of the *extension host*, inheriting `process.env` at spawn time (confirmed
 * against the `dotnet/vscode-csharp` source: `ProcessRunner`/`DotNetTestChannel`-style spawn calls
 * pass `env: process.env`). Setting `process.env.CustomAfterMicrosoftCommonTargets` here, as early
 * in extension activation as possible, is therefore the whole mechanism — genuinely unverified
 * without a live C# Dev Kit test run (issue #715 plan §7 risk #2): if some future C# Dev Kit
 * version spawns via a detached/isolated environment instead of inheriting the extension host's,
 * this injection silently does nothing (degrades to "no MTP outcomes," never a hard failure).
 */

export const CUSTOM_AFTER_MICROSOFT_COMMON_TARGETS_VARIABLE = 'CustomAfterMicrosoftCommonTargets';
const HOOK_GUID = 'a1d3c2f0-6b8e-4f2a-9c7d-3e5f8b1a4d6c';

const EXCLUDED_DIRECTORY_NAMES = new Set(['bin', 'obj', '.git', '.vs', '.vscode', 'node_modules']);

/**
 * Enumerates every `.csproj` under `root`, pruning build-output/VCS/editor/dependency directories.
 * Exported for testing (mirrors `MtpEphemeralInjection.EnumerateProjectFiles` on the VS side).
 */
export function enumerateProjectFiles(root: string): string[] {
  const found: string[] = [];
  const stack = [root];

  while (stack.length > 0) {
    const dir = stack.pop()!;
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      continue;
    }

    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!EXCLUDED_DIRECTORY_NAMES.has(entry.name)) {
          stack.push(path.join(dir, entry.name));
        }
      } else if (entry.isFile() && entry.name.toLowerCase().endsWith('.csproj')) {
        found.push(path.join(dir, entry.name));
      }
    }
  }

  return found;
}

/**
 * Writes the ephemeral `.targets` file declaring the reporter reference + builder hook, chain-
 * importing any pre-existing `CustomAfterMicrosoftCommonTargets` value so this injection composes
 * with whatever else (if anything) already uses that extensibility point. Exported for testing.
 */
export function writeTargetsFile(
  reporterDllPath: string,
  preExistingCustomAfterTargets?: string,
): string {
  const dir = path.join(
    os.tmpdir(),
    `reqnroll-mtp-inject-${crypto.randomBytes(16).toString('hex')}`,
  );
  fs.mkdirSync(dir, { recursive: true });
  const file = path.join(dir, 'Reqnroll.IdeSupport.TestReporter.MTP.g.targets');

  const chainImport = preExistingCustomAfterTargets
    ? `  <Import Project="${preExistingCustomAfterTargets}" Condition="Exists('${preExistingCustomAfterTargets}')" />\n`
    : '';

  const xml =
    '<Project>\n' +
    chainImport +
    '  <ItemGroup>\n' +
    '    <Reference Include="Reqnroll.IdeSupport.TestReporter.MTP">\n' +
    `      <HintPath>${reporterDllPath}</HintPath>\n` +
    '    </Reference>\n' +
    `    <TestingPlatformBuilderHook Include="${HOOK_GUID}">\n` +
    '      <DisplayName>Reqnroll.IdeSupport.TestReporter.MTP</DisplayName>\n' +
    '      <TypeFullName>Reqnroll.IdeSupport.TestReporter.MTP.TestingPlatformBuilderHook</TypeFullName>\n' +
    '    </TestingPlatformBuilderHook>\n' +
    '  </ItemGroup>\n' +
    '</Project>\n';

  fs.writeFileSync(file, xml, 'utf8');
  return file;
}

/**
 * Scans every workspace folder for an MTP-capable project and, if one is found, sets
 * `process.env.CustomAfterMicrosoftCommonTargets` for the remainder of this extension-host
 * process's lifetime. Returns `true` when it enabled injection, `false` otherwise (no MTP-capable
 * project found, or the bundled reporter is missing) — never throws.
 */
export async function tryEnableForWorkspace(
  workspaceFolderPaths: readonly string[],
  reporterDllPath: string,
): Promise<boolean> {
  try {
    if (!fs.existsSync(reporterDllPath)) {
      logWarn(
        `testOutcomes: bundled MTP reporter '${reporterDllPath}' not found; not enabling ephemeral injection.`,
      );
      return false;
    }

    const projectFiles = workspaceFolderPaths.flatMap((folder) => enumerateProjectFiles(folder));
    const mtpCapableFlags = await Promise.all(
      projectFiles.map((projectFile) => isMtpCapable(projectFile)),
    );
    const mtpCapable = mtpCapableFlags.some((capable) => capable);
    if (!mtpCapable) {
      logInfo(
        'testOutcomes: no MTP-capable project found in the workspace; not enabling ephemeral injection.',
      );
      return false;
    }

    const preExisting = process.env[CUSTOM_AFTER_MICROSOFT_COMMON_TARGETS_VARIABLE];
    const targetsFile = writeTargetsFile(reporterDllPath, preExisting);
    process.env[CUSTOM_AFTER_MICROSOFT_COMMON_TARGETS_VARIABLE] = targetsFile;
    logInfo(
      `testOutcomes: MTP ephemeral injection enabled for this session — ${CUSTOM_AFTER_MICROSOFT_COMMON_TARGETS_VARIABLE}=${targetsFile}`,
    );
    return true;
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    logWarn(`testOutcomes: MTP ephemeral injection failed — ${msg}`);
    return false;
  }
}

/**
 * Extension-activation entry point. Gated on the same `reqnroll.testOutcomes.enabled` opt-in as
 * `activateTestOutcomes` (this is part of the same feature — MTP-mode outcome reporting rather
 * than the VSTest-logger path) and must run as early in `activate()` as possible, before any test
 * run C# Dev Kit might launch could plausibly start. Never throws.
 */
export async function activateMtpEphemeralInjection(
  context: Pick<vscode.ExtensionContext, 'extensionMode' | 'extensionPath'>,
): Promise<void> {
  const enabled = vscode.workspace
    .getConfiguration('reqnroll')
    .get<boolean>('testOutcomes.enabled', false);
  if (!enabled) return;

  const reporterDirectory = resolveMtpReporterDirectory(context);
  if (!reporterDirectory) {
    logWarn(
      'testOutcomes: reqnroll.testOutcomes.enabled is true, but the bundled MTP reporter was not found; skipping ephemeral injection.',
    );
    return;
  }

  const workspaceFolderPaths = (vscode.workspace.workspaceFolders ?? []).map((f) => f.uri.fsPath);
  if (workspaceFolderPaths.length === 0) return;

  await tryEnableForWorkspace(
    workspaceFolderPaths,
    path.join(reporterDirectory, MTP_REPORTER_ASSEMBLY_FILE_NAME),
  );
}
